### PR TITLE
More structured suggestions for boxed trait objects instead of impl Trait on non-coerceable tail expressions

### LINKED ISSUE #69107: Detect when if/else is tail expression has mismatched types that could be a boxed trait
Given the following
```rust
struct S;
struct Y;

trait Trait {}

impl Trait for S {}
impl Trait for Y {}

fn baz() -> impl Trait {
    if true {
        S
    } else {
        Y
    }
}
```
[we currently emit](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=33b39a8ed7e34ca644fca8a23c5aa146)
```
error[E0308]: `if` and `else` have incompatible types
  --> src/main.rs:13:9
   |
10 | /     if true {
11 | |         S
   | |         - expected because of this
12 | |     } else {
13 | |         Y
   | |         ^ expected struct `S`, found struct `Y`
14 | |     }
   | |_____- `if` and `else` have incompatible types
```

[We should suggest changing](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=493cbd88403ec9abebf2c18dc3b7486a) the return type from `impl Trait` to `Box<dyn Trait>` and boxing the two branches. I think this is the last missing suggestion we can give involding `impl Trait` and `dyn Trait` returns.

Follow up to https://github.com/rust-lang/rust/issues/68110.

### PR DESCRIPTION
**Pull request by estebank**
When encountering a `match` or `if` as a tail expression where the
different arms do not have the same type *and* the return type of that
`fn` is an `impl Trait`, check whether those arms can implement `Trait`
and if so, suggest using boxed trait objects.

Use structured suggestion for `impl T` to `Box<dyn T>`.

Fix https://github.com/rust-lang/rust/issues/69107

### CODE REVIEW

**Comment by estebank** [comment]
I just want to see how much of a perf impact this has to see if it will even be worth it to clean up and land.

**Comment by estebank** [comment]
🤯 I am amazed at how little this affected perf. Expect a further iteration cleaning this up.

**Comment by estebank** [comment]
Screen Shot 2020-08-16 at 8 00 29 PM

File: src/librustc_typeck/check/_match.rs [hunk 1]
```diff
@@ -112,14 +115,74 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+
+            // When we have a `match` as a tail expression in a `fn` with a returned `impl Trait`
+            // we check if the different arms would work with boxed trait objects instead and
+            // provide a structured suggestion in that case.
+            let suggest_box = match (
+                orig_expected,
+                self.body_id
+                    .owner
+                    .to_def_id()
+                    .as_local()
```
Comment (lcnr) [inline src/librustc_typeck/check/_match.rs:127]: `HirId -> LocalDefId -> DefId -> Option<LocalDefId>`.

This seems incorrect to me, is there a reason you don't use `tcx.hir().opt_local_def_id(self.body_id)`?

Not really part of the PR, but why is `body_id` a `HirId` instead of a `BodyId` for `FnCtxt`?

File: src/librustc_typeck/check/_match.rs [hunk 2]
```diff
@@ -112,14 +115,74 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                    .and_then(|id| self.ret_coercion_impl_trait.map(|ty| (id, ty))),
+            ) {
+                (Expectation::ExpectHasType(expected), Some((id, ty)))
```
Comment (lcnr) [inline src/librustc_typeck/check/_match.rs:130]:
```suggestion
                (Expectation::ExpectHasType(expected), Some((body_id, ty)))
```
unclear what `id` means here

File: src/librustc_typeck/check/_match.rs [hunk 3]
```diff
@@ -266,6 +330,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         else_expr: &'tcx hir::Expr<'tcx>,
         then_ty: Ty<'tcx>,
         else_ty: Ty<'tcx>,
+        suggest_box: Option<Span>,
```
Comment (lcnr) [inline src/librustc_typeck/check/_match.rs:333]: `suggest_box` being an `Option` instead of a bool seems somewhat weird to me.

Not sure what's a better name here though :thinking: maybe `opt_suggest_box_span` or something like this.

File: src/test/ui/point-to-type-err-cause-on-impl-trait-return.stderr [hunk 4]
```diff
@@ -48,25 +54,38 @@ LL |         1u32
    |
    = note: to return `impl Trait`, all returned values must be of the same type
    = note: for information on `impl Trait`, see <https://doc.rust-lang.org/book/ch10-02-traits.html#returning-types-that-implement-traits>
-   = help: you can instead return a boxed trait object using `Box<dyn std::fmt::Display>`
    = note: for information on trait objects, see <https://doc.rust-lang.org/book/ch17-02-trait-objects.html#using-trait-objects-that-allow-for-values-of-different-types>
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/test/ui/point-to-type-err-cause-on-impl-trait-return.stderr:57]: Suggestions always come last. It's been in my backlog for a while to address that.

File: src/librustc_typeck/check/_match.rs (hunk 3 above)
Comment (estebank) [reply src/librustc_typeck/check/_match.rs:333]: It was originally a bool, which is why it is weird 😅

File: src/librustc_infer/infer/error_reporting/mod.rs [hunk 5]
```diff
@@ -613,11 +613,20 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                 ref prior_arms,
                 last_ty,
                 scrut_hir_id,
+                suggest_box,
+                arm_span,
                 ..
             }) => match source {
                 hir::MatchSource::IfLetDesugar { .. } => {
                     let msg = "`if let` arms have incompatible types";
```
[the remark this answers is not in the document]
Comment (estebank) [reply src/librustc_infer/infer/error_reporting/mod.rs:621]: No, I'll add one

File: src/librustc_typeck/check/_match.rs (hunk 1 above)
Comment (estebank) [reply src/librustc_typeck/check/_match.rs:127]: Using the proposed change causes us to lose the following suggestion:

```
error[E0308]: `if` and `else` have incompatible types
 --> fil3.rs:5:9
  |
2 | /     if false {
3 | |         0i32
  | |         ---- expected because of this
4 | |     } else {
5 | |         1u32 //~ ERROR `if` and `else` have incompatible types
  | |         ^^^^ expected `i32`, found `u32`
6 | |     }
  | |_____- `if` and `else` have incompatible types
  |
help: you could change the return type to be a boxed trait object
  |
1 | fn qux() -> Box<dyn std::fmt::Display> {
  |             ^^^^^^^                  ^
help: if you change the return type to expect trait objects, box the returned expressions
  |
3 |         Box::new(0i32)
4 |     } else {
5 |         Box::new(1u32) //~ ERROR `if` and `else` have incompatible types
  |
```

I think that the following is equivalent to what we had before:

```rust
            let body_parent_id = tcx.hir().get_parent_did(self.body_id);
            let opt_suggest_box_span = match (
                orig_expected,
                self.ret_coercion_impl_trait.map(|ty| (body_parent_id, ty))
```

Comment (estebank) [reply src/librustc_typeck/check/_match.rs:127]: Nevermind, just realized what a mess I got myself into that could be simplified quite a bit.

**Comment by estebank** [comment]
We still provide two errors for the case where a bare `dyn Trait` is the return type, but that "should be enough", at least for now:

```rust
fn qux() -> dyn std::fmt::Display {
    if false {
        0i32
    } else {
        1u32 //~ ERROR `if` and `else` have incompatible types
    }
}
```

```
error[E0308]: `if` and `else` have incompatible types
 --> fil3.rs:5:9
  |
2 | /     if false {
3 | |         0i32
  | |         ---- expected because of this
4 | |     } else {
5 | |         1u32 //~ ERROR `if` and `else` have incompatible types
  | |         ^^^^ expected `i32`, found `u32`
6 | |     }
  | |_____- `if` and `else` have incompatible types

error[E0746]: return type cannot have an unboxed trait object
 --> fil3.rs:1:13
  |
1 | fn qux() -> dyn std::fmt::Display {
  |             ^^^^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
  |
  = note: for information on trait objects, see <https://doc.rust-lang.org/book/ch17-02-trait-objects.html#using-trait-objects-that-allow-for-values-of-different-types>
  = note: if all the returned values were of the same type you could use `impl std::fmt::Display` as the return type
  = note: for information on `impl Trait`, see <https://doc.rust-lang.org/book/ch10-02-traits.html#returning-types-that-implement-traits>
  = note: you can create a new `enum` with a variant for each returned type
help: return a boxed trait object instead
  |
1 | fn qux() -> Box<dyn std::fmt::Display> {
2 |     if false {
3 |         Box::new(0i32)
4 |     } else {
5 |         Box::new(1u32) //~ ERROR `if` and `else` have incompatible types
  |

error: aborting due to 2 previous errors
```

**Review by lcnr — APPROVED** [review]
Didn't realize you fixed the merge conflicts :sweat_smile:

LGTM, but I don't feel completely confident in the changes to `fn check_match` so I will leave the final approval to someone else.

r? @varkor maybe :thinking:

File: compiler/rustc_typeck/src/check/_match.rs [hunk 6]
```diff
@@ -112,14 +115,70 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                                        },
+                                    },
+                                    constness,
+                                );
+                                let obl = Obligation::new(
+                                    o.cause.clone(),
+                                    self.param_env,
+                                    pred.to_predicate(self.infcx.tcx),
+                                );
+                                suggest_box &= self.infcx.predicate_must_hold_modulo_regions(&obl);
```
Comment (varkor) [inline compiler/rustc_typeck/src/check/_match.rs:154]: What case is this handling? (Maybe a comment would be helpful here.)

Comment (estebank) [reply compiler/rustc_typeck/src/check/_match.rs:153]: Added comment. All the obligations that come from a trait in an `impl Trait` return type need to be applicable to the tail expressions' types. If they do, then we can suggest to return a boxed trait object.
